alica_engine/include/engine/RunnableObjectNew.h

```diff
@@ -0,0 +1,131 @@
+    void traceRunCall();
+
+private:
+    TracingType _tracingType;
+    std::function<std::optional<std::string>()> _customTraceContextGetter;
+    std::unique_ptr<IAlicaTrace> _trace;
+    // True if the behaviour/plan's run method has already been logged in the trace
+    bool _runTraced;
+    IAlicaTraceFactory* _traceFactory;
+};
+
+/**
+ * The base class for BasicBehaviour and BasicPlan
+ */
+class RunnableObjectNew
+{
+protected:
+    using TracingType = TraceRunnableObject::TracingType;
+
+    RunnableObjectNew(IAlicaWorldModel* wm, const std::string& name = "");
+    virtual ~RunnableObjectNew() = default;
+    // virtual ~RunnableObjectNew() { std::cout << _name << " destructor" << std::endl; }
+
+    static constexpr int DEFAULT_MS_INTERVAL = 100;
+
+    virtual void doInit() = 0;
+    virtual void doRun() = 0;
+    virtual void doTerminate() = 0;
+
+    void setTracing(TracingType type, std::function<std::optional<std::string>()> customTraceContextGetter = {})
+    {
+        _runnableObjectTracer.setTracing(type, customTraceContextGetter);
+    }
+    const std::string& getName() { return _name; };
+    IAlicaTrace* getTrace() const { return _runnableObjectTracer.getTrace(); };
+
+    void sendLogMessage(int level, const std::string& message) const;
+    RunningPlan* getPlanContext() const { return _runningplanContext; }
+    const std::shared_ptr<Blackboard> getBlackboard() { return _blackboard; }
+    IAlicaWorldModel* getWorldModel() { return _wm; };
+
+    void start(RunningPlan* rp);
+    void stop();
+    void setEngine(AlicaEngine* engine) { _engine = engine; };
+    // Only plan will have these
+    void addKeyMapping(int64_t wrapperId, const KeyMapping* keyMapping);
+    void setInterval(int32_t msInterval) { _msInterval = AlicaTime::milliseconds(msInterval); };
+    bool getInheritBlackboard() const { return _blackboardBlueprint == nullptr; };
+    void setBlackboardBlueprint(const BlackboardBlueprint* blackboard) { _blackboardBlueprint = blackboard; }
+
+    AlicaEngine* _engine;
+
+private:
+    void setInput(const Blackboard* parent_bb, const KeyMapping* keyMapping);
+    void setOutput(Blackboard* parent_bb, const KeyMapping* keyMapping) const;
+    int64_t getParentWrapperId(RunningPlan* rt) const;
+    const KeyMapping* getKeyMapping(int64_t id) const { return _keyMappings.at(id); }
+    void setupBlackboard();
+    void scheduleRunCalls();
+    void stopRunCalls();
+    void cleanupBlackboard();
+    void runJob();
+
+    RunningPlan* _runningplanContext;
+    std::string _name;
+    AlicaTime _msInterval;
+    std::unique_ptr<IAlicaTimer> _activeRunTimer;
+    const BlackboardBlueprint* _blackboardBlueprint;
+    std::shared_ptr<Blackboard> _blackboard;
+    IAlicaWorldModel* _wm;
+    // Map from ConfAbstractPlanWrapper id to associated attachment
+    // Only plan will have these
+    std::unordered_map<int64_t, const KeyMapping*> _keyMappings;
+    TraceRunnableObject _runnableObjectTracer;
```

Comment:
```suggestion
    TraceRunnableObject _runnableObjectTrace;
```

Reply:
Its a tracer instance rather than trace. Just like NavigationTracer in LBC